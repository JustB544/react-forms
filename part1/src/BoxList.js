import React, {useState} from "react";
import Box from "./Box";
import {v4 as uuid} from 'uuid';
import NewBoxForm from "./NewBoxForm";

function BoxList(props) {
    const [boxes, setBoxes] = useState(props.initial || {});

    function deleteBox(id){
        setBoxes(({[id]: undefined, ...b}) => b);
    }

    function createBox({backgroundColor, height, width}){
        setBoxes(b => ({...b, [uuid()]: { backgroundColor, height, width }}));
    }

    return (
        <div>
            { Object.keys(boxes).map(b => <Box 
            key={b} 
            id={b}
            backgroundColor={boxes[b].backgroundColor} 
            height={boxes[b].height} 
            width={boxes[b].width} 
            deleteBox={evt => deleteBox(b)}/>) }
            <NewBoxForm createBox={(box, evt) => createBox(box)}/>
        </div>
    );
}

export default BoxList;