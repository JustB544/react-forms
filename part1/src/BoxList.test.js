import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";
import TEST_BOXES from "./_testcommon";

it("renders", () => {
    render(<BoxList />);
});
  
it("matches snapshot", () => {
    const {asFragment} = render(<BoxList initial={TEST_BOXES}/>);
    expect(asFragment()).toMatchSnapshot();
});

it("deletes a box", () => {
    const {container} = render(<BoxList initial={TEST_BOXES}/>);
    const box1 = container.querySelector("#box1 button");
    fireEvent.click(box1);
    expect(box1).not.toBeInTheDocument();
});