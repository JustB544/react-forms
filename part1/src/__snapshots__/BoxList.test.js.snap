// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="Box"
      id="box1"
    >
      <div
        style="background-color: rgb(0, 0, 0); height: 100px; width: 100px;"
      />
      <button>
        X
      </button>
    </div>
    <div
      class="Box"
      id="box2"
    >
      <div
        style="background-color: rgb(255, 255, 255); height: 1000px; width: 1000px;"
      />
      <button>
        X
      </button>
    </div>
    <form
      style="display: flex; flex-direction: column;"
    >
      <div>
        <label
          for="backgroundColor"
        >
          Background color: 
        </label>
        <input
          name="backgroundColor"
          type="color"
          value="#000000"
        />
      </div>
      <div>
        <label
          for="height"
        >
          Height: 
        </label>
        <input
          name="height"
          type="number"
          value="100"
        />
      </div>
      <div>
        <label
          for="width"
        >
          Width: 
        </label>
        <input
          name="width"
          type="number"
          value="100"
        />
      </div>
      <button>
        Create
      </button>
    </form>
  </div>
</DocumentFragment>
`;
